pkg/agent/controller/agent.go

```diff
@@ -228,94 +304,167 @@ func (a *Controller) serviceImportLister(transform func(si *mcsv1a1.ServiceImpor
+		// on the broker and delete it
+		return brokerServiceExport, false
 	}
 
-	obj, found, err := a.serviceSyncer.GetResource(svcExport.Name, svcExport.Namespace)
+	serviceObj, serviceExist, err := a.serviceSyncer.GetResource(localServiceExport.Name, localServiceExport.Namespace)
 	if err != nil {
-		// some other error. Log and requeue
-		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, corev1.ConditionUnknown, "ServiceRetrievalFailed",
-			fmt.Sprintf("Error retrieving the Service: %v", err))
-		klog.Errorf("Error retrieving Service (%s/%s): %v", svcExport.Namespace, svcExport.Name, err)
-
+		// some other error. Log and requeue to retry later
+		a.handleServiceExportTransformError(err, localServiceExport,
+			"ServiceRetrievalFailed", "Error retrieving the Service")
 		return nil, true
 	}
 
-	if !found {
-		klog.V(log.DEBUG).Infof("Service to be exported (%s/%s) doesn't exist", svcExport.Namespace, svcExport.Name)
-		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, corev1.ConditionFalse, serviceUnavailable,
-			"Service to be exported doesn't exist")
+	if !serviceExist {
+		// corresponding service not exist - requeue to retry later, until service is created
+		a.handleServiceExportTransformError(err, localServiceExport,
+			reasonServiceUnavailable, "Service to be exported does not exist, will not upload")
 
 		return nil, true
 	}
 
-	if op == syncer.Update && getLastExportConditionReason(svcExport) != serviceUnavailable {
+	// this is a status update of the local export and service exist.
+	// we want to continue only if the last status is "service unavailable"
+	// which means that service was previously not available and is now available
+	// so the export should be uploaded to the broker
+	if op == syncer.Update && getLastExportConditionReason(localServiceExport) != reasonServiceUnavailable {
 		return nil, false
 	}
 
-	svc := obj.(*corev1.Service)
+	svc := serviceObj.(*corev1.Service)
+
+	isSupportedServiceType := svc.Spec.Type == "" || svc.Spec.Type == corev1.ServiceTypeClusterIP
+	if !isSupportedServiceType {
+		a.handleServiceExportTransformError(err, localServiceExport,
+			"UnsupportedServiceType", fmt.Sprintf("Service of type %v not supported", svc.Spec.Type))
 
-	svcType, ok := getServiceImportType(svc)
+		return nil, false
+	}
 
-	if !ok {
-		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, corev1.ConditionFalse, invalidServiceType,
-			fmt.Sprintf("Service of type %v not supported", svc.Spec.Type))
-		klog.Errorf("Service type %q not supported", svc.Spec.Type)
+	exportSpec, err := mcs.NewExportSpec(svc, localServiceExport, a.clusterID)
+	if err != nil {
+		a.handleServiceExportTransformError(err, localServiceExport,
+			"ExportSpecCreationFailed", "Error creating service export spec")
 
 		return nil, false
 	}
 
-	serviceImport := a.newServiceImport(svcExport.Name, svcExport.Namespace)
+	err = exportSpec.MarshalObjectMeta(&brokerServiceExport.ObjectMeta)
```

Comment:
would `if err := function(); err != nil` be more consistent with the rest of the code?

Reply:
sorry, I really dont like the consolidated syntax.